fish-rust/src/trace.rs

```diff
@@ -0,0 +1,79 @@
+use crate::{
+    common::{escape_string, EscapeFlags},
+    ffi::{self, parser_t, wcharz_t},
+    wchar::{self, wstr, L},
+    wchar_ffi::WCharToFFI,
+};
+use std::sync::atomic;
+
+#[cxx::bridge]
+mod trace_ffi {
+    extern "C++" {
+        include!("wutil.h");
+        include!("parser.h");
+        type wcharz_t = super::wcharz_t;
+        type parser_t = super::parser_t;
+    }
+
+    extern "Rust" {
+        #[cxx_name = "trace_set_enabled"]
+        fn trace_set_enabled_ffi(do_enable: bool);
+        #[cxx_name = "trace_enabled"]
+        fn trace_enabled_ffi(parser: &parser_t) -> bool;
+        #[cxx_name = "trace_argv"]
+        fn trace_argv_ffi(parser: &parser_t, command: wcharz_t, args: &Vec<wcharz_t>);
+        #[cxx_name = "trace_if_enabled"]
+        fn trace_if_enabled_ffi(parser: &parser_t, command: wcharz_t);
+    }
+}
+
+static DO_TRACE: atomic::AtomicBool = atomic::AtomicBool::new(false);
+
+fn trace_set_enabled_ffi(do_enable: bool) {
+    DO_TRACE.store(do_enable, atomic::Ordering::Relaxed);
+}
+
+fn trace_enabled_ffi(parser: &parser_t) -> bool {
+    let ld = parser.ffi_libdata_pod_const();
+    if ld.suppress_fish_trace {
+        return false;
+    }
+    return DO_TRACE.load(atomic::Ordering::Relaxed);
+}
+
+/// Trace an "argv": a list of arguments where the first is the command.
+// Allow the `&Vec` parameter as this function only exists temporarily for the FFI
+#[allow(clippy::ptr_arg)]
+fn trace_argv_ffi(parser: &parser_t, command: wcharz_t, args: &Vec<wcharz_t>) {
+    let command: wchar::WString = command.into();
+    let args: Vec<wchar::WString> = args.iter().map(Into::into).collect();
+    let args_ref: Vec<&wstr> = args.iter().map(wchar::WString::as_utfstr).collect();
+    trace_argv(parser, command.as_utfstr(), &args_ref);
+}
+
+fn trace_argv(parser: &parser_t, command: &wstr, args: &[&wstr]) {
+    // Format into a string to prevent interleaving with flog in other threads.
+    // Add the + prefix.
+    let mut trace_text = L!("-").repeat(parser.blocks_size() - 1);
+    trace_text.push('>');
+
+    if !command.is_empty() {
+        trace_text.push(' ');
+        trace_text.push_utfstr(command);
+    }
+    for arg in args {
+        trace_text.push(' ');
+        trace_text.push_utfstr(&escape_string(
+            arg,
+            crate::common::EscapeStringStyle::Script(EscapeFlags::default()),
+        ));
+    }
+    trace_text.push('\n');
+    ffi::log_extra_to_flog_file(&trace_text.to_ffi());
+}
+
+fn trace_if_enabled_ffi(parser: &parser_t, command: wcharz_t) {
+    if trace_enabled_ffi(parser) {
+        trace_argv_ffi(parser, command, &Vec::new());
```

Comment:
You also define this function in C++ which means this one is unused. We should naturalize it to accomodate upcoming Rust callers
1. make it `pub`
2. drop `_ffi` (and drop it from the bridge)
3. call `trace_argv` instead of `trace_argv_ffi`
4. add the `argv` argument instead of always passing an empty slice. Rust does not have default arguments so we have to include it always

Reply:
Done